### DISCUSSION TITLE
better integer division

### DISCUSSION THREAD

**Discussion by LO-0116** [godotengine/godot-proposals#10659] [Scripting]
This shows a warning:
![image](https://github.com/user-attachments/assets/0bab8048-ebfa-47b2-a384-494f344d48a3)

This bloats the line-count and increase visual clutter:
![image](https://github.com/user-attachments/assets/420144dd-046d-41ef-a6de-2be66a042548)

This just looks terrible:
![image](https://github.com/user-attachments/assets/d36ba077-afa0-4cad-ba63-91848cac162b)

Disabling warning for integer division may cause bugs where integer division isn't desired.

Suggestion: adding a new operator like `//` that explicitly divides two integers without throwing a warning.

**Comment by AThousandShips** [comment]
See this rejected proposal:
* https://github.com/godotengine/godot-proposals/issues/1866

**Reply by LO-0116** [reply]
> See this rejected proposal:
> 
> * [Separate integer and real division operators in GDScript #1866](https://github.com/godotengine/godot-proposals/issues/1866)

what's this supposed to mean? A contentious, directionless discussion that doesn't answer why it'd be undesirable to add 1 more library function like `int_div()`?

**Reply by AThousandShips** [reply]
It's literally the same topic, and the same operator 🙃 I'm providing context for your question and for that this isn't a new conversation but has been discussed at length and rejected in the past

**Reply by LO-0116** [reply]
At length? Yes. Conclusively? Not that I can tell. Nobody in the thread actually explained why there shouldn't be a separate operator.

**Reply by AThousandShips** [reply]
That isn't required, the support wasn't enough and there wasn't sufficient arguments in *favor*, features aren't approved based on lack of arguments against them but arguments in favor, a feature isn't approved by default

But there *was* significant opposition from area maintainers, which is a very valid reason, so if you read through that thread you get the exact reason for the proposal being rejected

**Reply by LO-0116** [reply]
The latest comment in the thread is about 1 year old, opinions may have changed since then. 

I also can't see any objective reason for why a new function can't be added, only some arguments about how changing the existing operator may break backward compatibility. Honestly your response here seems like an extension of unwarranted emotional investment in the topic.

As for lack of support, it seems like you and some others are intent on making it impossible to gather support for proposals along this line, and again, I do not see an objective reason for why that should be.

**Reply by dalexeev** [reply]
> The latest comment in the thread is about 1 year old, opinions may have changed since then.

I don't think this behavior can be changed in 4.x for compatibility reasons, so there's little point in revisiting it now. 5.0 isn't planned yet, and 4.x is supposed to be the main development branch for many years.

> I also can't see any objective reason for why a new function can't be added, only some arguments about how changing the existing operator may break backward compatibility.

This only makes sense for `float` arguments (`int_div(a: float, b: float) -> int`), but you already can use casting the operands to `int` (`int(a) / int(b)`). In my opinion, new operators don't make sense, since the main operator still has confusing behavior, it would only complicate things.

Also note that some languages like Haskell ​​have two integer division operators (`div` and `quot`) and two remainder operators (`mod` and `rem`). We don't want to complicate the language that much, but perhaps `int_div()` / `intdiv()` makes sense as a `@GlobalScope` function, since `posmod()` already exists.

> As for lack of support, it seems like you and some others are intent on making it impossible to gather support for proposals along this line, and again, I do not see an objective reason for why that should be.

Please avoid such statements and implied accusations, assume positive intentions.

**Reply by girng** [reply]
> The latest comment in the thread is about 1 year old, opinions may have changed since then.
> 
> I also can't see any objective reason for why a new function can't be added, only some arguments about how changing the existing operator may break backward compatibility. Honestly your response here seems like an extension of unwarranted emotional investment in the topic.
> 
> As for lack of support, it seems like you and some others are intent on making it impossible to gather support for proposals along this line, and again, I do not see an objective reason for why that should be.

see reduz's post in that link brother. don't worry so much about why decisions are made, i try to adapt with what we have, and spend time with your dream game. i've been questioning issues about this engine for over 9 years, some stuff is just the way it is, and i learned it's ok. it's impossible for devs to please us all and conform to everything we want. that goes for any project tbh. 

for this particular issue, i'm on 2.1 but you can disable warnings in editor settings   on newer godot versions. i also like how it defaults to int division already, if i want a float i just add .0. super simple for me (reduz's  philosophy for gdscript too, be made simple and easy for game programmers to use)

**Comment by Mickeon** [comment]
> Disabling warning for integer division may cause bugs where integer division isn't desired.

[Ideally](https://github.com/godotengine/godot/pull/69012), all of the warning settings would be moved to the Editor Settings.

You may also be interested in https://github.com/godotengine/godot/pull/76020 to disable the warning for the entire script.
